src/databricks/labs/ucx/progress/tables.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import replace
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.hive_metastore.tables import Table
+from databricks.labs.ucx.hive_metastore.table_migration_status import TableMigrationIndex
+from databricks.labs.ucx.hive_metastore.ownership import TableOwnership
+from databricks.labs.ucx.progress.grants import GrantProgressEncoder
+from databricks.labs.ucx.progress.history import ProgressEncoder
+from databricks.labs.ucx.progress.install import Historical
+
+
+class TableProgressEncoder(ProgressEncoder[Table]):
+    """Encoder class:Table to class:History.
+
+    A progress failure for a table means:
+    - the table is not migrated yet
+    - the associated grants have a failure
+    """
+
+    def __init__(  # pylint disable=too-many-arguments
+        self,
+        sql_backend: SqlBackend,
+        ownership: TableOwnership,
+        table_migration_index: TableMigrationIndex,
+        grant_progress_encoder: GrantProgressEncoder,
+        run_id: int,
+        workspace_id: int,
+        catalog: str,
+        schema: str = "multiworkspace",
+        table: str = "historical",
+    ) -> None:
+        super().__init__(
+            sql_backend,
+            ownership,
+            Table,
+            run_id,
+            workspace_id,
+            catalog,
+            schema,
+            table,
+        )
+        self._table_migration_index = table_migration_index
+        self._grant_progress_encoder = grant_progress_encoder
```

Comment:
@nfx: Please provide input on linking the grants to the table.

The progress encoder currently only has an append to history method, thus not possibly to get the historical grants (with failures) out of it easily.

Also, I am leaning to ignore the grants as some companies might chose to not migrate it and I doubt if we should show a table as not being migrated if a `READ_METADATA` or `DENY` grant cannot be mapped to UC

Reply:
I left grants out-of-scope for the `TableProgressEncoder` and documented why I did so, please review